[FEAT] 카카오페이 결제 기능 구현

chaechaen opened the pull request:
## #⃣ 연관된 이슈
- Closes #45 

## 1. 📝 작업 내용

> 작업 내용이 많고 헷갈려서 그냥 블로그 정리 느낌으로 쭉 쓰겠습니다..ㅎ

> 참고: [카카오페이 단건 결제 공식 문서](https://developers.kakaopay.com/docs/payment/online/single-payment)

### 카카오페이 결제 준비 기능 구현 (`/kakaoPay/ready`)
- MemberTicket 정보 기반으로 결제 요청 생성
- 카카오페이 서버에 WebClient를 통해 비동기 요청 전송
- `next_redirect_pc_url`(결제창 URL)과 `tid`를 응답으로 반환
- 응답 받은 `tid`는 MemberTicket에 저장 후 승인 요청 시 재사용하도록 함

### 카카오페이 결제 승인 기능 구현 (`/kakaoPay/approve`)
- 카카오페이 결제 완료 후 자동 호출되는 redirect 엔드포인트
- 전달된 `pg_token`, DB에서 조회한 `tid`, 그리고 `partner_order_id`, `partner_user_id`, `cid`를 포함하여 승인 요청을 카카오 서버에 전송
- 결제 승인 성공 시 기존 MemberTicketService에 있던 로직 적용 (옮김)
  - MemberTicket의 상태를 `PENDING → RESERVED`로 변경
  - 공연 회차(`AmateurRounds`)의 남은 티켓 수 감소, 해당 공연의 누적 판매 티켓 수 증가
  - 위 로직은 `KakaoPayBusinessService`로 분리

### 승인 API의 인증 정보 처리 개선
- 카카오 redirect 요청에는 인증 토큰이 포함되지 않아 기존의 `@AuthenticationPrincipal` 사용 시 에러 발생
- 이를 해결하기 위해 컨트롤러의 `approve()`에서 인증 객체를 제거하고,  서비스단에서 `partner_order_id`로 MemberTicket을 조회한 뒤, 관련된 Member를 추적하여 사용

### WebClient 사용
- 카카오페이 서버와 통신할 때 WebClient를 사용해 비동기 요청을 처리함
- JPA 호출은 blocking 방식이라 non-blocking 방식인 WebClient랑 같이 쓰면 블로킹 이슈가 생길 수 있다고 해서   `Mono.fromCallable(...).subscribeOn(Schedulers.boundedElastic())`로 감싸서 별도 스레드에서 따로 실행되게 처리함
- WebClient는 non-blocking으로 유지하고, DB 조회만 따로 분리해서 실행되도록 리액티브 구조 구성

    
## 4. **📸 스크린샷 (선택)**
- /kakaoPay/approve는 카카오 redirect에 의해 자동 호출되므로 스웨거상에선 사용하지 말 것!!
      <img width="923" alt="image" src="https://github.com/user-attachments/assets/cfcd37cf-1185-4c1f-9d8d-bba82a52716b" />
- 승인 성공 시 이렇게 JSON 응답 확인 가능
      <img width="266" alt="image" src="https://github.com/user-attachments/assets/181a27f1-9d52-4b95-941c-834bb2b294b1" />

- 테스트 결제 화면 예시
  <p align="left">
    <img src="https://github.com/user-attachments/assets/e8328c1a-4650-419b-8e3b-4b0e8bd5585d" width="150" height="300"/>
    <img src="https://github.com/user-attachments/assets/f6cd33a0-39c7-41c6-b28e-a3ffbb8704ec" width="250" height="300"/>
  </p>


    
## 5. 테스트 방법
- Swagger에서 POST /kakaoPay/ready 호출
- `ticketId` 파라미터 입력하고, 응답으로 받은 `next_redirect_pc_url`을 브라우저에서 접속하면 카카오페이 결제 페이지 QR코드가 나옴
- QR 인증 후 결제 완료
- 결제 완료 시 카카오가 자동으로 백엔드의 `/kakaoPay/approve?partner_order_id=xxx&pg_token=xxx`으로 redirect 요청
- 해당 요청이 `approve()` API를 자동 호출하여 결제 승인 처리
- 승인 성공 시 JSON 응답 확인 (결제 승인 정보)
    - !! 참고로 pg_token은 1회용이므로, 같은 결제에 대해 approve API를 중복 호출하면 실패합니다 !!

## 6. 추가 구현해야 할 사항
- 아직 개발 환경에서 테스트하느라 url을 localhost:8080 사용 중입니다.. 추후 실제 도메인 변경하겠습니다.
- 그리고 현재는 카카오페이 테스트용 CID(TC0ONETIME)를 사용하고 있습니다. 실제 가맹점 코드를 발급 받으면 CID 변경하겠습니다.

chaechaen committed:
feat: 카카오페이 결제 기능 구현

chaechaen committed:
feat: 카카오결제 필요 dto 및 주석 추가

chaechaen committed:
fix: 카카오결제 승인 요청 구조 개선

chaechaen committed:
fix: 멤버티켓 수량 감소 및 상태 변경 로직 삭제 (카카오 서비스단으로 옮김)

chaechaen committed:
merge: origin develop pulled

chaechaen committed:
merge: origin develop pulled

chaechaen committed:
refactor: WebClient + JPA 블로킹 분리 및 리액티브 구조 개선

chaechaen committed:
fix: 카카오 redirect 요청 시 인증 정보 없이 처리되도록 approve API 수정

ddhi7 commented on src/main/java/cc/backend/ticket/service/MemberTicketServiceImpl.java:
```diff
@@ -68,17 +68,18 @@ public MemberTicketCreateResponseDTO createTicket(Long amateurRoundId, Long amat
+//        amateurTicket.getAmateurShow().increaseSoldTicket(requestDTO.getQuantity());
 
         //티켓 예매 알림 이벤트 생성
         eventPublisher.publishEvent(new TicketReservationEvent(ticket.getAmateurTicket().getAmateurShow(), ticket.getAmateurTicket(), member));
```
티켓 생성시 이벤트쪽에서 오류나고있는 것 같습니다..!

ddhi7 commented on src/main/java/cc/backend/kakaoPay/service/KakaoPayBusinessService.java:
```diff
@@ -0,0 +1,35 @@
+            throw new GeneralException(ErrorStatus.MEMBER_TICKET_ALREADY_RESERVED);
+        }
+
+        ticket.updateTid(tid);
```
Tid ready에서랑 여기서 두번 update 하고있는 것 같아요

chaechaen replied:
KakaoPayService의 `/ready` 로직에서 이미 tid가 저장되므로, 결제 승인 단계에서는 중복 저장을 방지하기 위해 해당 업데이트 부분을 삭제했습니다!

ddhi7 commented on src/main/java/cc/backend/kakaoPay/service/KakaoPayBusinessService.java:
```diff
@@ -0,0 +1,35 @@
+
+        ticket.updateTid(tid);
+        ticket.updateReservationStatus(ReservationStatus.RESERVED);
+        ticket.getAmateurRound().decreaseTotalTicket(ticket.getQuantity());
```
결제완료 후에 재고수량을 단순 감소시키고있는데 이렇게하면 '동시성 문제'가 발생할 수 있을 것 같아요! 락 걸어 놓고 결제완료 후에도 다시 재고 수량 체크 한 후에 감소시켜야할 것 같습니당
```java
      // 비관적 락
        AmateurRounds round = amateurRoundsRepository.findByIdWithLock(ticket.getAmateurRound().getId())
                .orElseThrow(() -> new GeneralException(ErrorStatus.ROUND_NOT_FOUND));
        
        // 재고 확인
        if (round.getTotalTicket() < ticket.getQuantity()) {
            throw new GeneralException(ErrorStatus.MEMBER_TICKET_STOCK);
        }

//차감
decrease
        
```

ddhi7 replied:
예를 들어, 티켓 1개가 남은 상황에서 1분간격으로 두명이 티켓 생성(티켓 예약 요청) ->  처음 티켓 생성한 사람이 결제가 끝나지않아서 재고가 업데이트 되지않았을 테니 두번 째 사람이 티켓 생성가능 -> 두 사람 모두 결제하게되면 동시성 문제 발생

chaechaen replied:
넵 말씀해주신 사항 확인했고 아래와 같이 수정했습니다!

### 문제

남은 재고가 1장일 때, 동시에 여러 사용자가 결제 승인을 요청하면 모든 요청이 동시에 처리되며 재고가 음수가 되거나, 여러 티켓이 `RESERVED` 상태로 바뀌는 동시성 문제가 발생할 수 있습니다. 그리고 초기 구조에서는 `/ready`에서 단순히 재고 수량 조회 후 조건만 검사한 뒤, 실제 재고 차감은 `/approve` 단계에서 이루어졌기 때문에 사용자는 이미 카카오페이에서 결제 완료 알림을 받은 상태지만, 우리 서버에서 재고 부족으로 예외가 발생하는 정합성 문제가 있었습니다.

### 해결

1. **WebClient 비동기 호출 제거**
    - 기존에는 `.subscribeOn()`으로 작성했었지만, 이 경우 WebClient 호출 이후 로직이 별도의 스레드에서 실행되면서 트랜잭션 컨텍스트가 전파되지 않고, 락이 제대로 작동하지 않는 문제가 발생할 수 있었습니다.
    - 따라서 동기식 블로킹 호출 `.block()`만 사용하도록 수정하여, 외부 API 호출 시에도 같은 스레드 내에서 실행되고 트랜잭션 컨텍스트가 유지되도록 했습니다.

2. **조건부 원자적 업데이트 적용**
    - `AmateurRoundsRepository`에 DB 원자성을 이용해 재고 확인과 차감을 한 번에 처리하는 쿼리인 `decreaseStock` 메소드를 추가하여, 동시에 여러 요청이 발생해도 DB 레벨에서 딱 1개 요청만 성공시키고 나머지는 실패 처리되도록 했습니다.
    - 이 로직은 전부 `/ready` 단계에서 이루어지고, `/approve` 단계에서는 상태 확정만 하도록 했습니다. 이를 통해 TOCTOU 문제를 해결했습니다.

### JUnit 동시성 테스트

3가지 시나리오에 대한 동시성 테스트를 작성하고 성공한 것을 확인했습니다.

- 시나리오 1: 재고(1개)보다 많은 수량(2개) 요청 시, `/ready` 단계에서 즉시 실패
- 시나리오 2: 마지막 재고(1개)를 두고 3명이 동시에 요청 시, 1명만 성공하고 나머지 2명은 실패
- 시나리오 3 (TOCTOU 방지): 재고가 4개일 때 3명이 각각 2개씩 요청 시, 정확히 2명(총 4개)만 성공하고 마지막 1명은 실패

ddhi7 commented on src/main/java/cc/backend/ticket/service/MemberTicketServiceImpl.java:
```diff
@@ -68,17 +68,18 @@ public MemberTicketCreateResponseDTO createTicket(Long amateurRoundId, Long amat
                 .cancelAvailableUntil(round.getPerformanceDateTime().minusDays(1).withHour(17))
                 .totalPrice(totalPrice)
-                .reservationStatus(ReservationStatus.RESERVED)
+                .reservationStatus(ReservationStatus.PENDING) // PENDING 으로
```
PENDING 으로 바뀌어서 티켓 생성이 티켓 예약 요청 역할을 하게됨에따라 티켓 발급 로직이 추가로 필요할 것 가타요 
티켓 생성(티켓 예약) -> 결제 (RESERVED로 상태변환)-> 티켓 발급

ddhi7 replied:
좀 더 찾아보니 간단한 사용 여부와 예매번호는 티켓 발급로직이 꼭 필요한 건 아니라구하네욥

chaechaen replied:
추가로 결제 실패 처리 관련해선 결제 준비 단계에서 `cancelUrl`, `failUrl`을 지정해두었습니다.
- 사용자가 결제를 15분 동안 진행하지 않는 경우 → `failUrl`로 리디렉트됨
- 인증 중 취소하거나, 결제 창에서 취소 버튼을 누른 경우 → `cancelUrl`로 리디렉트

그리고 사용자가 `/ready`까지는 성공했는데 결제를 완료하지 않고 이탈하면 이미 ready 단계에서 재고가 차감된 상태가 되어버리는 문제가 발생할 수 있어서 재고 복구에 대한 로직도 추가하겠습니다..!

ddhi7 reviewed (commented):
너무 고생하셨습니다!! 혹시 결제 오류가 나게될 경우에는 어떻게되나요?_?

chaechaen committed:
refactor: 중복된 tid 업데이트 제거

chaechaen committed:
fix: 결제 승인 로직 락 적용하여 동시성 문제 개선

chaechaen committed:
test: 결제 승인 race condition 시나리오 테스트 코드 추가

chaechaen committed:
refactor: 결제 재고 차감 로직 정합성 문제 개선

chaechaen committed:
test: 결제 재고 로직에 대한 시나리오 테스트 추가

chaechaen committed:
test: 결제 재고 선점 동시성 테스트 추가

chaechaen committed:
fix: 비관적 락 제거 후 조건 기반 재고 차감 방식으로 변경

chaechaen committed:
Merge branch 'develop' of https://github.com/SeeATheater/CC_Backend into feat/payment

chaechaen commented:
추가로 결제 실패 처리 관련해선 결제 준비 단계에서 `cancelUrl`, `failUrl`을 지정해두었습니다.
- 사용자가 결제를 15분 동안 진행하지 않는 경우 → `failUrl`로 리디렉트됨
- 인증 중 취소하거나, 결제 창에서 취소 버튼을 누른 경우 → `cancelUrl`로 리디렉트

그리고 사용자가 `/ready`까지는 성공했는데 결제를 완료하지 않고 이탈하면 이미 ready 단계에서 재고가 차감된 상태가 되어버리는 문제가 발생할 수 있어서 재고 복구에 대한 로직도 추가하겠습니다..!

ddhi7 reviewed (approved):
와 테스트코드까지.. 고생하셨습니다~~

chaechaen committed:
Merge branch 'develop' of https://github.com/SeeATheater/CC_Backend into feat/payment

chaechaen committed:
feat: 만료된 PENDING 티켓 자동 처리 스케줄러 추가

- 15분 이상 결제 대기 상태인 티켓을 주기적으로 조회하여 EXPIRED 상태로 변경
- EXPIRED 처리 시 해당 회차의 재고 복구 로직 포함
- 5분 간격으로 실행되도록 스케줄러 설정

chaechaen committed:
feat: 재고 상태 검증 로직 추가

chaechaen committed:
feat: SecurityConfig에 카카오페이 관련 API 접근 허용

chaechaen committed:
refactor: 카카오페이 승인/취소/실패 URL을 application.yml로 이동
